### PR TITLE
detection transforms and their utils added

### PR DESCRIPTION
**Pull request by shaydeci**
Second PR as part of integrating Yolox to SG, that holds detection transforms and their utilities.

### CODE REVIEW

File: src/super_gradients/training/transforms/transforms.py [hunk 1]
```diff
@@ -334,4 +341,359 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+            labels = np.hstack((box_labels, cls_labels))
+            origin_labels = np.vstack((origin_labels, labels))
+            origin_img = origin_img.astype(np.float32)
+            origin_img = 0.5 * origin_img + 0.5 * padded_cropped_img.astype(np.float32)
+
+            sample["image"], sample["target"] = origin_img.astype(np.uint8), origin_labels
+        return sample
+
+
+class YoloxTrainPreprocessFN(DetectionTransform):
```
Comment (ofrimasad) [inline src/super_gradients/training/transforms/transforms.py:607]: it is a bit weird to define it as a DetectionTransform. it's the Yolox dataset get_item function...

File: src/super_gradients/training/transforms/transforms.py [hunk 2]
```diff
@@ -334,4 +341,359 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+        padded_labels = np.zeros((self.max_labels, 5))
+        padded_labels[range(len(targets_t))[: self.max_labels]] = targets_t[
+                                                                  : self.max_labels
+                                                                  ]
+        padded_labels = np.ascontiguousarray(padded_labels, dtype=np.float32)
+        sample["image"], sample["target"] = image_t, padded_labels
+        return sample
+
+
+class YoloxValPreprocessFN(DetectionTransform):
```
Comment (ofrimasad) [inline src/super_gradients/training/transforms/transforms.py:675]: same here. looks like this should be a a part of the yolox dataset code

File: src/super_gradients/training/utils/detection_utils.py [hunk 3]
```diff
@@ -1164,3 +1168,355 @@ def num_anchors(self) -> int:
 
     def __repr__(self):
         return f"anchors_list: {self.__anchors_list} strides: {self.__strides}"
+
+
+def get_yolox_datadir():
```
Comment (ofrimasad) [inline src/super_gradients/training/utils/detection_utils.py:1173]: how is that a util? 
we can add getenv to super_gradients.training.utils.utils.get_param (but not in this PR)

**Review by ofrimasad — COMMENTED** [review]
my comments inline

File: src/super_gradients/training/utils/detection_utils.py (hunk 3 above)
Comment (shaydeci) [reply src/super_gradients/training/utils/detection_utils.py]: removed

File: src/super_gradients/training/utils/detection_utils.py [hunk 4]
```diff
@@ -1164,3 +1168,355 @@ def num_anchors(self) -> int:
...
+    :param w_max: (int) width border.
+    :param h_max: (int) height border
+    :return: modified bbox (np.array)
+    """
+    bbox[:, 0::2] = np.clip(bbox[:, 0::2] * scale_ratio + padw, 0, w_max)
+    bbox[:, 1::2] = np.clip(bbox[:, 1::2] * scale_ratio + padh, 0, h_max)
+    return bbox
+
+
+class YoloXCollateFN:
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/utils/detection_utils.py:1267]: added

File: src/super_gradients/training/transforms/transforms.py [hunk 5]
```diff
@@ -334,4 +341,722 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+            cp_img[: int(img.shape[0] * cp_scale_ratio), : int(img.shape[1] * cp_scale_ratio)] = resized_img
+
+            cp_img = cv2.resize(
+                cp_img,
+                (int(cp_img.shape[1] * jit_factor), int(cp_img.shape[0] * jit_factor)),
+            )
+            cp_scale_ratio *= jit_factor
+
+            if FLIP:
+                cp_img = cp_img[:, ::-1, :]
```
Comment (ofrimasad) [inline src/super_gradients/training/transforms/transforms.py:566]: why not use _mirror() here as well?

**Review by ofrimasad — COMMENTED** [review]
two unresolved from previous review and 3 new comments

File: src/super_gradients/training/transforms/transforms.py [hunk 6]
```diff
@@ -334,4 +341,722 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+            boxes = targets[:, 1:]
+            labels = targets[:, 0]
+        else:
+            boxes = targets[:, :4]
+            labels = targets[:, 4]
+
+        if convert2cxcy:
+            boxes = xyxy2cxcywh(boxes)
+        elif convert2xyxy:
+            boxes = cxcywh2xyxy(boxes)
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py:798]: No need to convert anything regarding the coordinates (for example just changing the order and normalising)

File: src/super_gradients/training/transforms/transforms.py [hunk 7]
```diff
@@ -334,4 +341,359 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+
+
+class Mosaic(DetectionTransform):
+    """
+    Mosaic detection transform
+    
+    Attributes:
+        input_dim: (tuple) input dimension.
+        prob: (float) probability of applying mosaic.
+        enable_mosaic: (bool) whether to apply mosaic at all (regardless of prob) (default=True).
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py:385]: As we discussed- we need it to turn on/off. Added to init.

File: src/super_gradients/training/transforms/transforms.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py]: Removed it as now it was dismanteled into different transforms.

File: src/super_gradients/training/transforms/transforms.py [hunk 8]
```diff
@@ -334,4 +341,722 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+        self.enable_mixup = False
+
+    def __call__(self, sample: dict):
+        if self.enable_mixup and random.random() < self.prob:
+            origin_img, origin_labels = sample["image"], sample["target"]
+            cp_sample = sample["additional_samples"][0]
+            img, cp_labels = cp_sample["image"], cp_sample["target"]
+
+            jit_factor = random.uniform(*self.mixup_scale)
+            FLIP = random.uniform(0, 1) > 0.5
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py]: Removed.

File: src/super_gradients/training/transforms/transforms.py (hunk 5 above)
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py]: Fixed.

**Review by ofrimasad — APPROVED** [review]
minor comment and then LGTM

File: src/super_gradients/training/transforms/transforms.py [hunk 9]
```diff
@@ -334,4 +341,651 @@ def _validate_fill_values_arguments(fill_mask: int, fill_image: Union[int, Tuple
...
+
+class DetectionMixup(DetectionTransform):
+    """
+    Mixup detection transform
+
+    Attributes:
+        input_dim: (tuple) input dimension.
+        mixup_scale: (tuple) scale range for the additional loaded image for mixup.
+        prob: (float) probability of applying mixup.
+        enable_mixup: (bool) whether to apply mixup at all (regardless of prob) (default=True).
```
[the remark this answers is not in the document]
Comment (shaydeci) [reply src/super_gradients/training/transforms/transforms.py:522]: done

**Pull request merged**